### ISSUE TITLE
Link header for HTML and JSON-LD

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#204]
If we remove support for context extraction from HTML, we might want to consider broadening the Link Header to be able to specify the context location, which might help @danbri. We also might want to revisit whether it can bu used for JSON-LD files as wee, as it’s currently restricted to just plain JSON.

**Comment by pchampin** [comment]
Let me rephrase. What you are suggesting is to *link* a JSON-LD context from an HTML page rather than embedding it. This would be done using [RFC5988](https://tools.ietf.org/html/rfc5988) HTTP headers, so
* it would not require the JSON-LD processor to parse the HTML, and
* it would not require the server to perform "active" content negotiation.

Am I getting this correctly?

**Comment by dlongley** [comment]
If I understand this, this approach sounds like it wouldn't be that functionally different from following a redirect. So, a JSON-LD processor would hit a URL, get back something that isn't `application/ld+json` but that has a link header in it that points at a JSON-LD context. We already do something very similar for JSON resources that have a context elsewhere ... so it doesn't seem to me like that much of a stretch to support this if it actually addresses important use cases.

cc: @msporny  -- thoughts?

**Comment by gkellogg** [comment]
The thought I expressed on the call was to re-use the Link header but add semantics so that if the response when fetching a context is _not_ `application/ld+json` and it includes a Link header with `rel="...#context"` a processor would load the context from the specified location. (creates another opportunity for infinite recursion, which would need to be handled).

**Comment by BigBlueHat** [comment]
Let's be sure to distinguish between the HTTP header conversation and the HTML-level `<link>` tag approach. They have very different requirements in terms of both processing and deployment.

Also @gkellogg's concerns about infinite recursion would likely apply to both scenarios.

**Comment by azaroth42** [comment]
Minutes of previous call: https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-07-19-json-ld#section5-1

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-08-09-json-ld#section4-3).

- No actions or resolutions
<details><summary><i class="differentiate">View the transcript</i></summary>
Link header for HTML and JSON-LD<br/>
Benjamin Young: See Issue 204<br/>
Gregg Kellogg: this was trying to rely to danbri’s concern about relying on content negotiation<br/>
… the best solution I came up with was to rely on a link HTTP header<br/>
… A JSON-LD processor expecting a context at a given URL would check for this header,<br/>
… and follow the link if any.<br/>
Ivan Herman: other people do that, with rel="manifest"<br/>
… the Web Publication WG does something similar with rel="publication"<br/>
… we can bikeshed about what rel value we want. But this makes sense.<br/>
Ruben Taelman: link headers are not always possible on static websites, e.g. github pages.<br/>
Benjamin Young: just to be clear, the link header we are talking about is a HTTP header, not HTML $MD_CODE$<br/>
… It is not as common as content negotiation on static sites.<br/>
… So we should specify that JSON-LD processors should first try content-negotiation,<br/>
… if it fails, it should try HTML (which implies embedding a HTML parser)<br/>
Dave Longley: +1 to link headers, -1 to involving HTML processing at all w/context processing as it breaks the ecosystem<br/>
Benjamin Young: then it should check the link header.<br/>
Pierre-Antoine Champin: I think this solution makes sense<br/>
… whether content negotiation has priority or not, I think this would be nice to have<br/>
… the way I saw it this was a way to get rid of a dependency on HTTP<br/>
Gregg Kellogg: this is all because the desire of schema.org (and may be others)<br/>
… to avoid doing content negotiation.<br/>
… Whenever we can’t do conneg nor HTTP headers,<br/>
… we are back to parsing HTML.<br/>
Dave Longley: I would be in favor of the link HTTP header solution,<br/>
… even if it does not solve all problems.<br/>
… I am against the reliance on HTML parser. This would break the ecosystem.<br/>
… Also, we should not make decision on the features that github pages have today.<br/>
… That may change in the future.<br/>
Benjamin Young: (a cosmetic argument that I missed)<br/>
Ruben Taelman: I don’t see the benefit of the link header over conneg;<br/>
… if a host supports one, it should supports the other, right?<br/>
Dave Longley: some static sites can do link headers.<br/>
… It requires less processing on the server side than conneg.<br/>
Pierre-Antoine Champin: +1<br/>
</details>

**Comment by gkellogg** [comment]
# Proposal
In [LoadDocumentCallback](https://www.w3.org/TR/json-ld11-api/#loaddocumentcallback) (which defines how _all_ JSON-LD processors should handle document retrieval) after step 4:

1) If the retrieved resource's `Content-Type` is not application/json or any media type with a +json suffix as defined in [RFC6839] and and the response has an HTTP `Link` Header [RFC8259] using the `alternate` link relation with type `application/ld+json`, set _url_ to the associated `href` and restart the algorithm at step 2, ensuring that the resulting resource is of type `application/ld+json`.

So, in the case that an HTML resource is returned with a link header such as the following:

```HTTP
Link: <https://schema.org/docs/jsonldcontext.jsonld>; rel="alternate"; type="application/ld+json"
```

the document loader would see that the `Content-Type` was text/html and that a Link header was included which directed to `https://schema.org/docs/jsonldcontext.jsonld` and the document loader would essentially redirect to that resource, which MUST then be of type `application/ld+json`

We'd also need a section in the syntax document near [Interpreting JSON as JSON-LD](https://www.w3.org/TR/json-ld11/#interpreting-json-as-json-ld) to describe this behavior and note it's utility for static websites. This would also allow us to remove the need to process embedded HTML for context files, as this handles the documentation requirement.

We could consider a different link relation, but "alternate" seems appropriate, when coupled with type.

cc/ @danbri

**Comment by iherman** [comment]
Minor:

> and and the response

one 'and' too much...

Otherwise +1 for me

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-08-16-json-ld#section4-2).

- `RESOLVED:  merge api#134 and syntax#212, and close syntax#204`
<details><summary><i class="differentiate">View the transcript</i></summary>
Benjamin Young: Link header: preemptive conneg.<br/>
Benjamin Young: See Syntax issue #204<br/>
Benjamin Young: See API PR #134<br/>
Benjamin Young: See Syntax PR #212<br/>
Benjamin Young: 2 PRs are ready to go, right?<br/>
Gregg Kellogg: Yes. Open question for base during redirect.<br/>
… If you retrieve something non-json. Doc loader will detect rel-alternative links, and redirect to that resource.<br/>
… I think the original requested document should be preserved as base URL.<br/>
… Just like 303 redirect semantics. In 302, this would not be the case.<br/>
Rob Sanderson: JS will silently follow redirects and give you the results. So you don’t know which redirects got you there.<br/>
Gregg Kellogg: JS should correctly implement redirect semantics.<br/>
Dave Longley: +1 to using 303 semantics, so on, +1 to gregg’s position.<br/>
Gregg Kellogg: You have processing state to keep track on this.<br/>
Proposed resolution: merge api#134 and syntax#212, and close syntax#204 (Ivan Herman)<br/>
Rob Sanderson: +1<br/>
Dave Longley: +1<br/>
Benjamin Young: +1<br/>
Rob Sanderson: :shipit:<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Ruben Taelman: +0.5<br/>
Resolution #2: merge api#134 and syntax#212, and close syntax#204<br/>
</details>

**Comment by gkellogg** [comment]
This particular issue relates to #211, while #212 is for #172.
